### PR TITLE
Support unknown types (backend)

### LINKED ISSUE #3024: Support unknown Postgres types
Mathesar currently (erroniously) presumes that each Postgres data type it sees will be known via the relevant enums ([see comment](https://github.com/centerofci/mathesar/issues/3024#issuecomment-1632812393)) and, when an unknown type is encountered, Mathesar panics ([see bug report](https://github.com/centerofci/mathesar/issues/2709)). We want Mathesar to handle unknown types gracefully.

### PR DESCRIPTION
**Pull request by dmos62**
Fixes #3024

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: db/types/operations/convert.py [hunk 1]
```diff
@@ -18,12 +19,11 @@ def get_db_type_enum_from_id(db_type_id):
         try:
             return MathesarCustomType(db_type_id)
         except ValueError:
+            if db_type_id == UnknownType().id:
+                return UnknownType()
```
Comment (dmos62) [inline db/types/operations/convert.py:23]: This change especially, but also the other change in this file might seem funny. The motivation was to get `get_db_type_enum_from_id` and `get_db_type_enum_from_class` to work with unknown types, and it turns out that most of the rest of the code fell into place.

**Comment by kgodey** [comment]
Does this also fix related issues #2709, #2959, and #3007? Or do those need additional work?

**Comment by silentninja** [comment]
@kgodey It fixes partially fixes #2709 by not throwing up an error when an unknown type is encountered. In order to fix #2709, #2959 and #3007 we need to add support for those types(citext, point) instead of considering them as "Unknown Type"

**Comment by dmos62** [comment]
@kgodey I'll expand on @silentninja's answer by saying that this fixes "unknown types" (types we've not enumerated) breaking the column and record APIs (and maybe reflection as well, I don't recall). There are two complimentary directions to go from here:

1. have backend report the actual type identifiers of unknown types: this PR groups all of them under the umbrella identifier `__unknown__`; I didn't do this in this PR, because it's potentially-complex and it's unnecessary in order to downgrade this problem from a highest-priority blocker to a medium-priority UX concern;
2. have backend know things about some of these types; for example, even if Mathesar doesn't crash when `citext` is used, it still doesn't know that it's a text type, so it's not able to provide the right subset of filters or the right type casting choices (actually we disallow casting from unknown types).

**Comment by kgodey** [comment]
Thanks @dmos62, sounds good. I do agree not crashing is the most immediate thing to fix, and 1 and 2 are secondary (but helpful to have).

**Review by silentninja — APPROVED** [review]
Looks good!

On a side note, it took some time for me to understand what was going on, mostly because there is an intermediary SQLAlchemy type involved in the mapping process. We definitely need to re-visit the type mapping and see if this can be made simple.

**Comment by mathemancer** [comment]
@dmos62 In [the docs](https://www.postgresql.org/docs/13/catalog-pg-type.html), they give the following table:

Code | Category
-- | --
A | Array types
B | Boolean types
C | Composite types
D | Date/time types
E | Enum types
G | Geometric types
I | Network address types
N | Numeric types
P | Pseudo-types
R | Range types
S | String types
T | Timespan types
U | User-defined types
V | Bit-string types
X | unknown type

These codes are associated with all types in the DB, and you can query the `pg_type` table to get this info. I was thinking we could use these categories to simplify some of our logic, and also to do a better job of types which are unknown to Mathesar, but _are_ known to the DB.

As an example, suppose the `mathesar_types.email` type was unknown to Mathesar. This type is implemented as a domain over `TEXT`. This means that the `pg_type` table puts it into the `S` (for String) category, which we could at least use to give the front end a clue about how to handle input for the type and so on, even without knowing anything about it in the back end.

**Pull request merged**